doc/neps/nep-0021-advanced-indexing.rst

```diff
@@ -0,0 +1,537 @@
+  multi-dimensional indexing such as ``arr[[slice(None), 2]]``.
+  This creates some inconsistencies and thus the indexing attributes
+  should only allow plain python tuples for this purpose.
+  (Whether or not this should be the case for plain indexing is a
+  different issue.)
+
+* The new attributes should not do the getitem to implement setitem
+  branch, since it is a cludge and not useful for vectorized
+  indexing. (not implemented yet)
+
+
+Open Questions
+~~~~~~~~~~~~~~
+
+* The names ``oindex`` and ``vindex`` are just suggestions at the time of
+  writing this, another name NumPy has used for something like ``oindex``
+  is ``np.ix_``. See also below.
+
+* It would be possible to limit the use of boolean indices in ``vindex``,
+  assuming that they are rare and to some degree special. Alternatively,
+  boolean indices could be broadcasted as in legacy fancy indexing in
+  principle.
+
+* ``oindex`` and ``vindex`` could always return copies, even when no array
+  operation occurs. One argument for allowing a view return is that this way
+  ``oindex`` can be used as a general index replacement.
+  However, there is one argument for returning copies. It is possible for
+  ``arr.vindex[array_scalar, ...]``, where ``array_scalar`` should be
+  a 0-D array but is not, since 0-D arrays tend to be converted.
+  Copying always "fixes" this possible inconsistency.
+
+* The final state to morph plain indexing in is not fixed in this PEP.
+  It is for example possible that `arr[index]`` will be equivalent to
+  ``arr.oindex`` at some point in the future.
+  Since such a change will take years, it seems unnecessary to make
+  specific decisions at this time.
+
+* The proposed changes to plain indexing could be postponed indefinitely or
+  not taken in order to not break or force major fixes to existing code bases.
+
+* Should there be/is it possible to have a mechanism to not allow the new
+  indexing attributes for subclasses unless specifically implemented?
+
+* Should we try to warn users about the special indexing attributes
+  not being implemented? If a subclass has its own indexing, inheriting
+  it from ndarray should be wrong.
+ 
+
+
+Alternative Names
+~~~~~~~~~~~~~~~~~
+
+Possible names suggested (more suggestions will be added).
+
+==============  ======== ======= ============
+**Orthogonal**  oindex   oix
+**Vectorized**  vindex   vix
+**Legacy**      l/findex         legacy_index
+==============  ======== ======= ============
+
+
+Subclasses
+~~~~~~~~~~
+
+Subclasses are a bit problematic in the light of these changes. There are
+some possible solutions for this. For most subclasses (those which do not
+provide ``__getitem__`` or ``__setitem__``) the special attributes should
+just work. Subclasses that *do* provide it must be updated accordingly
+and should preferably not subclass working versions of these attributes.
+
+All subclasses will inherit the attributes, however, it seems possible
```

Comment:
Probably, the that was hanging me was that getting masked arrays right was damn annoying, plus we also need to have a way for things like memmap. I am not sure why a kwarg is super nice if that PEP is accepted, we get a mix of different meanings for kwargs, so trying to not use kwargs because of the PEP might be a point?

Reply:
Yes, on closer reading of the PEP, my suggestion would require the `kwarg` strategy; https://www.python.org/dev/peps/pep-0472/#strategy-kwargs-argument

Now it doesn't look like the PEP is actually going anywhere...

Anyway, my overall sense is that this should not hold up the NEP, or, indeed, the implementation.

Reply:
We could potentially call `self.__getitem__(indices, index_method='outer')` directly. We just can't write `self[indices, index_method='outer']`.

Reply:
That is indeed what I had in mind, and I thought the PEP might eventually make that possible as well, but it turns out that is true only for one of its options. Anyway, at some level that is a separate concern. The main advantage of the `index_method` approach is that subclasses would fail automatically (and that one could change the default at some point).

Reply:
I misread the PEP, my gut feeling is, if `arr[indices, index_method='outer']` would work, that would not be nice. Unless you want to use it instead of the `arr.oindex[indices]` syntax maybe. Simply because `index_method='outer'` could also mean get me "outer" from the axis "index_method"?

Reply:
Yes, it would preclude having an axis called `index_method`!

Anyway, right now we cannot use keywords inside `[...]` at all, so that part is not so relevant. The main question is whether it is worth solving the subclass issue by having `__getitem__` acquire an extra argument that we use internally (and `oindex` and `vindex` just being shortcuts to it).

I should add that I don't really have an opinion one way or another; I do think it is fine to expect subclasses to pay attention and adjust to new `ndarray` features (we just have to try our best not to break existing features).